File: part1_sec2/javascript/type_system_demo.js
// JavaScript: dynamic, weakly-typed
// Variables can be reassigned to any type, and mixed-type operations
// are silently coerced instead of raising an error.

function describe(value) {
    console.log(`${JSON.stringify(value)} is of type ${typeof value}`);
}

let age = 25;
describe(age);
age = "twenty-five";       // legal: dynamic typing allows re-binding
describe(age);

let score = 10;
let bonus = "5";
let total = score + bonus;  // number + string: '+' coerces to string concatenation
console.log("score + bonus =", total, "(type:", typeof total + ")");

console.log("Correct way:", score + Number(bonus));

File: part1_sec2/python/type_system_demo.py
# Python: dynamic, strongly-typed
# Same variable name can be re-bound to any type, but mixed-type
# operations that don't make sense are rejected at run time.

def describe(value):
    print(f"{value!r} is of type {type(value).__name__}")

age = 25
describe(age)
age = "twenty-five"   # legal: dynamic typing allows re-binding to a new type
describe(age)

score = 10
bonus = "5"
try:
    total = score + bonus   # int + str: no implicit coercion in Python
except TypeError as e:
    print("TypeError:", e)

print("Correct way:", score + int(bonus))
